Support extra channel output

lvandeve opened the pull request:
Allows outputting extra channels to individual single-channel buffers,
similar to the main image output

Adds two new functions to the API: JxlDecoderExtraChannelBufferSize
and JxlDecoderSetExtraChannelBuffer

jonsneyers commented:
For the CMYK case, it might make more sense to get output as 4-channel CMYK (like we do with RGBA now) than to get a "main image" that is CMY and then a separate extra channel for the K.

For the RGBG case (Bayer data), the same is probably true: you likely want to get all channels together.

For now, this is good enough, but perhaps it would make sense to have other 4-channel output paths than RGBA, in case there is some other extra channel that is not alpha but still interesting to interleave with the main RGB data (kBlack or kCFA spring to mind, but possibly kDepth too?). This should be relatively easy to implement (just selecting a different extra channel to act as the "alpha channel" in the RGBA code path).

The general case of interleaved N-channel output (for N>4) is probably not worth doing and just planar access to the extra channels is good enough in those cases (e.g. when CMYKA is used or multiple spot colors).

veluca93 commented on lib/jxl/decode.cc:
```diff
@@ -438,10 +438,15 @@ struct JxlDecoderStruct {
   JxlPixelFormat image_out_format;
 
+  // For extra channels
+  std::vector<bool> extra_channel_wanted;  // Empty if none, reset each frame
```
Can we avoid vector<bool> here?

lvandeve replied:
You mean for all or only the bool one? What do you propose instead?

veluca93 replied:
uh, github ate my answer :P `vector<bool>`

lvandeve replied:
Done, removed

veluca93 commented on lib/jxl/decode_test.cc:
```diff
@@ -2415,6 +2416,87 @@ TEST(DecodeTest, AnimationTestStreaming) {
+
+  EXPECT_EQ(JXL_DEC_FULL_IMAGE, JxlDecoderProcessInput(dec));
+
+  // After the full image is gotten, JxlDecoderProcessInput should return
```
"is gotten"?

lvandeve replied:
Changed to "was output"

veluca93 reviewed (dismissed):
Code LGTM

deymo commented on lib/jxl/decode.cc:
```diff
@@ -882,29 +889,28 @@ static size_t GetStride(const JxlDecoder* dec, const JxlPixelFormat& format,
-                                             void* out_image, size_t out_size,
-                                             JxlImageOutCallback out_callback,
-                                             void* out_opaque) {
+static JxlDecoderStatus ConvertImageInternal(
```
Can we have documentation of what's this function doing? It does quite different things depending on `want_extra_channel`; if it is true we get the extra channel only; but if it is false we get the color image. Maybe worth it to  have a different function for the "want_extra_channel==true" case, since it uses a different set of parameters, but for example do we need the RGBBuffer if want_extra_channel is true? We should assert that num_channels==1 in that case too since that should have been handled before.

lvandeve replied:
Fixed for external_image with https://github.com/libjxl/libjxl/pull/325, and documented for this internal function. I believe the internal function is simple enough to not split it into 3 functions (RGB(A) case, extra channel case, and the shared implementation), and documentation is added.

deymo commented on lib/jxl/dec_external_image.cc:
```diff
@@ -281,6 +290,8 @@ Status ConvertToExternal(const jxl::ImageBundle& ib, size_t bits_per_sample,
   Image3F temp_color, unpremul;
   const ImageF* alpha = ib.HasAlpha() ? &ib.alpha() : nullptr;
   ImageF temp_alpha;
+  const ImageF* extra = &ib.extra_channels()[extra_channel_index];
```
`extra = want_extra_channel ? ... : nullptr;`  extra_channel_index is not relevant if want_extra_channel is false and even if you specify that it must be 0 it could be possible that extra_channels is empty.

lvandeve replied:
No longer an issue after https://github.com/libjxl/libjxl/pull/325

deymo commented on lib/jxl/dec_external_image.cc:
```diff
@@ -408,6 +429,9 @@ Status ConvertToExternal(const jxl::ImageBundle& ib, size_t bits_per_sample,
             if (want_alpha) {
               row_in[c++] = ib.HasAlpha() ? alpha->Row(y) : ones.Row(0);
             }
+            if (want_extra_channel) {
```
I know this is correct, but it would make it more obvious what's going on if you do `if (want_extra_channel) { row_in[c++] = extra->Row(y); } else { ... set color and alpha ... }`. I think the fact that this logic is repeated 3 times merits a refactor of this code, for example you can split setting up which Planes need to be converted and separately the code that actually converts them, transposes them, etc.

lvandeve replied:
No longer an issue after https://github.com/libjxl/libjxl/pull/325

deymo commented on lib/jxl/dec_external_image.cc:
```diff
@@ -303,20 +314,27 @@ Status ConvertToExternal(const jxl::ImageBundle& ib, size_t bits_per_sample,
     color = &unpremul;
```
Do we need to do this unpremul processing in the want_extra_channel case? I think this ends up not being used so we don't need to spend time here. This shows the problems of reusing this function to do something different when a bool parameter is true.

lvandeve replied:
No longer an issue after https://github.com/libjxl/libjxl/pull/338

deymo commented on lib/jxl/dec_external_image.h:
```diff
@@ -35,6 +35,7 @@ namespace jxl {
 // xsize and ysize.
 Status ConvertToExternal(const jxl::ImageBundle& ib, size_t bits_per_sample,
```
Please document what's the meaning of want_extra_channel and what are the restrictions about calling this function with want_extra_channel=true.

lvandeve replied:
Done

deymo commented on lib/jxl/decode.cc:
```diff
@@ -1073,7 +1079,8 @@ JxlDecoderStatus JxlDecoderProcessInternal(JxlDecoder* dec, const uint8_t* in,
       if (want_preview) {
         if (dec->preview_out_buffer) {
           JxlDecoderStatus status = ConvertImageInternal(
```
Are there extra channels in the preview?

lvandeve replied:
Yes, the same image metadata applies to it. Outputting those can be added in a later merge request with a function such as `JxlDecoderSetExtraChannelPreviewOutBuffer`

deymo reviewed (commented):
Please add a call to these function to the djxl_fuzzer and run it for a bit, I'm pretty sure it would have detected indexing the empty vector.

lvandeve commented on lib/include/jxl/decode.h:
```diff
@@ -827,6 +827,57 @@ JXL_EXPORT JxlDecoderStatus
+ * @param dec decoder object
+ * @param format format of the pixels. Object owned by user and its contents
+ * are copied internally. The num_channels value is ignored and is always
+ * treated to be 1.
```
Done

lvandeve commented on lib/include/jxl/decode.h:
```diff
@@ -827,6 +827,57 @@ JXL_EXPORT JxlDecoderStatus
+ */
+JXL_EXPORT JxlDecoderStatus
+JxlDecoderSetExtraChannelBuffer(JxlDecoder* dec, const JxlPixelFormat* format,
+                                void* buffer, size_t size, size_t index);
```
Done

lvandeve commented on lib/jxl/decode.cc:
```diff
@@ -438,10 +438,14 @@ struct JxlDecoderStruct {
   JxlPixelFormat image_out_format;
 
+  // For extra channels. Empty of none, and they are reset each frame
+  std::vector<JxlPixelFormat> extra_channel_format;
```
Done, using a struct now

lvandeve commented:
> Please add a call to these function to the djxl_fuzzer and run it for a bit, I'm pretty sure it would have detected indexing the empty vector.

Added to fuzzer. Not sure which empty vector you mean though.

deymo commented:
> Added to fuzzer. Not sure which empty vector you mean though.

`  const ImageF* extra = &ib.extra_channels()[extra_channel_index];`

lvandeve commented:
> > Added to fuzzer. Not sure which empty vector you mean though.
> 
> ` const ImageF* extra = &ib.extra_channels()[extra_channel_index];`

I see, that code is gone. The API returns error if you try to get external image with smaller index. The fuzzer now 

> > Added to fuzzer. Not sure which empty vector you mean though.
> 
> ` const ImageF* extra = &ib.extra_channels()[extra_channel_index];`

I see, that code is gone. The API checks if you give a valid index, and it's now also fuzzed.

I also added back error checking for grayscale vs 3 channels in JxlDecoderImageOutBufferSize and JxlDecoderPreviewOutBufferSize, which the fuzzer expects. I had removed this due to reusing the code for extra channels where 1 channel does not mean grayscale, however the fuzzer expects it, and it makes sense that the "Size" functions do the same error checks as the "set buffer" functions.

PTAL

deymo commented on tools/djxl_fuzzer.cc:
```diff
@@ -334,6 +337,22 @@ bool DecodeJpegXl(const uint8_t* jxl, size_t size, size_t max_pixels,
       }
 
+      if (info.num_extra_channels > 0) {
+        size_t ec_index = info.num_extra_channels - 1;
```
Probably a good idea to make the requested index a random number % (info.num_extra_channels). Or better % (info.num_extra_channles+1) and use one value to not request extra channels even if there are extra channels.

lvandeve replied:
Done

lvandeve commented on lib/jxl/decode.cc:
```diff
@@ -438,10 +444,12 @@ struct JxlDecoderStruct {
   JxlPixelFormat preview_out_format;
   JxlPixelFormat image_out_format;
 
+  // For extra channels. Empty of none, and they are reset each frame
```
Was typo for "if", made the sentence longer

lvandeve commented on lib/jxl/dec_external_image.h:
```diff
@@ -40,6 +40,12 @@ Status ConvertToExternal(const jxl::ImageBundle& ib, size_t bits_per_sample,
                          size_t out_size, JxlImageOutCallback out_callback,
                          void* out_opaque, jxl::Orientation undo_orientation);
 
+Status ConvertToExternal(const jxl::ImageF& channel, size_t bits_per_sample,
```
Done

deymo reviewed (approved):
Thanks for the updates!
